Passing array literal as a parameter to a function that expects a pointer results in a C-level const qualifier discard

- **Kit version:** ad8de26ee216
- **OS/Platform:** Mac OS with clang (likely all platforms though)

A "passing 'const int [3]' to parameter of type 'int *' discards qualifiers" warning is emitted when compiling this code:

```kit
function fun(a: Ptr[Int]) {
}

function main() {
    fun([1, 2, 3]);
}
```

Compiler output:
```
% kitc test.kit
(...)
[2018-11-04 13:25:18.162932] /usr/bin/cc -Ibuild/include -c build/lib/kit_main.c -o build/obj/kit_main.o -D_GNU_SOURCE -D_BSD_SOURCE -D_DEFAULT_SOURCE '-std=c99' -pedantic -O3 -Os -U__BLOCKS__ -Wno-expansion-to-defined -Wno-gnu-zero-variadic-macro-arguments
build/lib/kit_main.c:6:19: warning: passing 'const int [3]' to parameter of type 'int *' discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]
        test__fun((const int [3]) {(int) 1, (int) 2, (int) 3});
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
build/include/main.h:222:22: note: passing argument to parameter 'a' here
void test__fun(int * a);
                     ^
1 warning generated.
(...)
```

I don't think there's a way to express `Ptr[const Int]` at the moment, is there?

--- a/src/Kit/Ast/BasicType.hs
+++ b/src/Kit/Ast/BasicType.hs
@@ -18,6 +18,7 @@ data BasicType
   | CPtr BasicType
   | BasicTypeConst BasicType
   | BasicTypeCInt
+  | BasicTypeCUint
   | BasicTypeCChar
   | BasicTypeCSize
   | BasicTypeVoid

--- a/src/Kit/Ast/BasicType.hs
+++ b/src/Kit/Ast/BasicType.hs
@@ -16,6 +16,7 @@ import Kit.Str
 data BasicType
   = CArray BasicType (Maybe Int)
   | CPtr BasicType
+  | BasicTypeConst BasicType
   | BasicTypeCInt
   | BasicTypeCChar
   | BasicTypeCSize
@@ -88,6 +89,7 @@ instance Show BasicType where
   show (CArray t Nothing) = show t ++ "[]"
   show (CPtr BasicTypeCChar) = "CString"
   show (CPtr t) = "Ptr[" ++ show t ++ "]"
+  show (BasicTypeConst t) = "Const[" ++ show t ++ "]"
   show (BasicTypeVoid) = "Void"
   show (BasicTypeBool) = "Bool"
   show (BasicTypeCChar) = "Char"
@@ -120,9 +122,11 @@ instance Show BasicType where
 type BasicArgs = [(Str, BasicType)]
 
 typeIsIntegral :: BasicType -> Bool
-typeIsIntegral (BasicTypeInt  _) = True
-typeIsIntegral (BasicTypeUint _) = True
-typeIsIntegral BasicTypeCChar    = True
-typeIsIntegral BasicTypeCInt     = True
-typeIsIntegral BasicTypeCSize    = True
-typeIsIntegral _                 = False
+typeIsIntegral (BasicTypeInt  _)  = True
+typeIsIntegral (BasicTypeUint _)  = True
+typeIsIntegral BasicTypeCChar     = True
+typeIsIntegral BasicTypeCInt      = True
+typeIsIntegral BasicTypeCUint     = True
+typeIsIntegral BasicTypeCSize     = True
+typeIsIntegral (BasicTypeConst t) = typeIsIntegral t
+typeIsIntegral _                  = False

--- a/src/Kit/Ast/Types.hs
+++ b/src/Kit/Ast/Types.hs
@@ -108,6 +108,8 @@ pattern TypeBox :: TypePath -> [ConcreteType] -> ConcreteType
 pattern TypeBox tp params = TypeInstance (["kit", "common"], "Box") [TypeTraitConstraint (tp, params)]
 pattern TypeArray :: ConcreteType -> Int -> ConcreteType
 pattern TypeArray t n = TypeInstance (["kit", "common"], "CArray") [t, ConstantType (IntValue n)]
+pattern TypeConst :: ConcreteType -> ConcreteType
+pattern TypeConst t = TypeInstance (["kit", "common"], "Const") [t]
 
 instance Show ConcreteType where
   show (TypePtr t) = "Ptr[" ++ show t ++ "]"

--- a/src/Kit/Compiler/Generators/C/CExpr.hs
+++ b/src/Kit/Compiler/Generators/C/CExpr.hs
@@ -246,7 +246,7 @@ transpileExpr (IrCArrLiteral values x) = u $ CCompoundLit
       )
       : snd t
     )
-  arrDecl = u $ CDecl ((CTypeQual $ u CConstQual) : (map CTypeSpec typeSpec))
+  arrDecl = u $ CDecl (map CTypeSpec typeSpec)
                       [(Just cdeclr, Nothing, Nothing)]
    where
     (typeSpec, derivedDeclr) = arrT

--- a/src/Kit/Compiler/Generators/C/CExpr.hs
+++ b/src/Kit/Compiler/Generators/C/CExpr.hs
@@ -54,7 +54,7 @@ cIdent x =
 
 cDecl :: BasicType -> Maybe TypePath -> Maybe (CInitializer NodeInfo) -> CDecl
 cDecl t ident body = u
-  $ CDecl (map CTypeSpec typeSpec) [(Just cdeclr, body, Nothing)]
+  $ CDecl typeSpec [(Just cdeclr, body, Nothing)]
  where
   (typeSpec, derivedDeclr) = ctype t
   cdeclr                   = u $ CDeclr
@@ -74,29 +74,33 @@ cpos p x = x $ mkNodeInfoOnlyPos $ Language.C.Data.Position.position
   (startCol p)
   Nothing
 
-ctype :: BasicType -> ([CTypeSpec], [CDerivedDeclr])
-ctype BasicTypeVoid    = ([u CVoidType], [])
-ctype BasicTypeBool    = ([u CBoolType], [])
-ctype (BasicTypeCChar) = ([u CCharType], [])
-ctype (BasicTypeCInt ) = ([u CIntType], [])
-ctype (BasicTypeCSize) = ([u $ CTypeDef (internalIdent "size_t")], [])
+ctype :: BasicType -> ([CDeclSpec], [CDerivedDeclr])
+ctype (BasicTypeConst t) =
+  let (a, b) = ctype t in ((CTypeQual $ u CConstQual) : a, b)
+ctype BasicTypeVoid    = ([CTypeSpec $ u CVoidType], [])
+ctype BasicTypeBool    = ([CTypeSpec $ u CBoolType], [])
+ctype (BasicTypeCChar) = ([CTypeSpec $ u CCharType], [])
+ctype (BasicTypeCInt ) = ([CTypeSpec $ u CIntType], [])
+ctype (BasicTypeCUint) = ([CTypeSpec $ u CIntType], [])
+ctype (BasicTypeCSize) =
+  ([CTypeSpec $ u $ CTypeDef (internalIdent "size_t")], [])
 ctype (BasicTypeInt n) =
-  ([u $ CTypeDef (internalIdent $ "int" ++ show n ++ "_t")], [])
+  ([CTypeSpec $ u $ CTypeDef (internalIdent $ "int" ++ show n ++ "_t")], [])
 ctype (BasicTypeUint n) =
-  ([u $ CTypeDef (internalIdent $ "uint" ++ show n ++ "_t")], [])
-ctype (BasicTypeFloat 32) = ([u CFloatType], [])
-ctype (BasicTypeFloat 64) = ([u CDoubleType], [])
+  ([CTypeSpec $ u $ CTypeDef (internalIdent $ "uint" ++ show n ++ "_t")], [])
+ctype (BasicTypeFloat 32) = ([CTypeSpec $ u CFloatType], [])
+ctype (BasicTypeFloat 64) = ([CTypeSpec $ u CDoubleType], [])
 ctype (BasicTypeFloat _ ) = undefined
 ctype (BasicTypeStruct name) =
-  ( [ u $ CSUType $ u $ CStruct CStructTag
-                                (Just $ cIdent $ mangleName name)
-                                Nothing
-                                []
+  ( [ CTypeSpec $ u $ CSUType $ u $ CStruct CStructTag
+                                            (Just $ cIdent $ mangleName name)
+                                            Nothing
+                                            []
     ]
   , []
   )
 ctype (BasicTypeAnonStruct Nothing args) =
-  ( [ u $ CSUType $ u $ CStruct
+  ( [ CTypeSpec $ u $ CSUType $ u $ CStruct
         CStructTag
         Nothing
         (Just
@@ -109,15 +113,15 @@ ctype (BasicTypeAnonStruct Nothing args) =
   , []
   )
 ctype (BasicTypeUnion name) =
-  ( [ u $ CSUType $ u $ CStruct CUnionTag
-                                (Just $ cIdent $ mangleName name)
-                                Nothing
-                                []
+  ( [ CTypeSpec $ u $ CSUType $ u $ CStruct CUnionTag
+                                            (Just $ cIdent $ mangleName name)
+                                            Nothing
+                                            []
     ]
   , []
   )
 ctype (BasicTypeAnonUnion Nothing args) =
-  ( [ u $ CSUType $ u $ CStruct
+  ( [ CTypeSpec $ u $ CSUType $ u $ CStruct
         CUnionTag
         Nothing
         (Just
@@ -130,7 +134,12 @@ ctype (BasicTypeAnonUnion Nothing args) =
   , []
   )
 ctype (BasicTypeSimpleEnum name) =
-  ([u $ CEnumType $ u $ CEnum (Just (cIdent $ mangleName name)) Nothing []], [])
+  ( [ CTypeSpec $ u $ CEnumType $ u $ CEnum (Just (cIdent $ mangleName name))
+                                            Nothing
+                                            []
+    ]
+  , []
+  )
 -- ctype (BasicTypeAnonEnum variants) =
 --   ( [ u $ CEnumType $ u $ CEnum Nothing
 --                                 (Just )
@@ -140,7 +149,13 @@ ctype (BasicTypeSimpleEnum name) =
 --   )
 ctype (BasicTypeComplexEnum name) = ctype (BasicTypeStruct name)
 ctype (BasicTypeTuple name t) =
-  ([u $ CSUType $ u $ CStruct CStructTag (Just (cIdent name)) Nothing []], [])
+  ( [ CTypeSpec $ u $ CSUType $ u $ CStruct CStructTag
+                                            (Just (cIdent name))
+                                            Nothing
+                                            []
+    ]
+  , []
+  )
 ctype (CPtr x) = (fst t, (u $ CPtrDeclr []) : snd t) where t = ctype x
 ctype (BasicTypeFunction rt args var) =
   let (rta, rtb) = ctype rt
@@ -152,9 +167,10 @@ ctype (BasicTypeFunction rt args var) =
         )
       : rtb
       )
-ctype (BasicTypeCFile) = ([u $ CTypeDef (internalIdent $ "FILE")], [])
+ctype (BasicTypeCFile) =
+  ([CTypeSpec $ u $ CTypeDef (internalIdent $ "FILE")], [])
 ctype (BasicTypeAnonEnum Nothing variants)
-  = ( [ u $ CEnumType $ u $ CEnum
+  = ( [ CTypeSpec $ u $ CEnumType $ u $ CEnum
           Nothing
           (Just
             [ (internalIdent $ s_unpack variant, Nothing)
@@ -178,10 +194,14 @@ ctype (CArray x s) =
     : snd t
   )
   where t = ctype x
-ctype (BasicTypeAnonStruct (Just x) _) = ([u $ CTypeDef $ cIdent x], [])
-ctype (BasicTypeAnonUnion  (Just x) _) = ([u $ CTypeDef $ cIdent x], [])
-ctype (BasicTypeAnonEnum   (Just x) _) = ([u $ CTypeDef $ cIdent x], [])
-ctype (BasicTypeUnknown              ) = undefined
+ctype (BasicTypeAnonStruct (Just x) _) =
+  ([CTypeSpec $ u $ CTypeDef $ cIdent x], [])
+ctype (BasicTypeAnonUnion (Just x) _) =
+  ([CTypeSpec $ u $ CTypeDef $ cIdent x], [])
+ctype (BasicTypeAnonEnum (Just x) _) =
+  ([CTypeSpec $ u $ CTypeDef $ cIdent x], [])
+ctype (BasicTypeUnknown) = undefined
+-- ctype (t) = throwk $ InternalError ("Unhandled ctype: " ++ show t) Nothing
 
 intFlags f = foldr (\f acc -> setFlag f acc) noFlags f
 
@@ -246,8 +266,8 @@ transpileExpr (IrCArrLiteral values x) = u $ CCompoundLit
       )
       : snd t
     )
-  arrDecl = u $ CDecl (map CTypeSpec typeSpec)
-                      [(Just cdeclr, Nothing, Nothing)]
+  arrDecl = u
+    $ CDecl typeSpec [(Just cdeclr, Nothing, Nothing)]
    where
     (typeSpec, derivedDeclr) = arrT
     cdeclr                   = u $ CDeclr Nothing derivedDeclr Nothing []

--- a/src/Kit/Compiler/Generators/C/CFun.hs
+++ b/src/Kit/Compiler/Generators/C/CFun.hs
@@ -10,7 +10,7 @@ import Kit.Str
 cfunDecl :: FunctionDefinition IrExpr BasicType -> CDecl
 cfunDecl f@(FunctionDefinition { functionName = name, functionType = rt, functionArgs = args, functionVarargs = varargs })
   = u $ CDecl
-    (map CTypeSpec typeSpec)
+    typeSpec
     [ ( Just $ u $ CDeclr
         (Just $ cIdent $ mangleName name)
         ((u $ CFunDeclr (Right (map cfunArg args, varargs)) []) : derivedDeclr)
@@ -25,7 +25,7 @@ cfunDecl f@(FunctionDefinition { functionName = name, functionType = rt, functio
 cfunDef :: FunctionDefinition IrExpr BasicType -> CFunDef
 cfunDef f@(FunctionDefinition { functionName = name, functionType = rt, functionArgs = args, functionVarargs = varargs, functionBody = Just body })
   = u $ CFunDef
-    (map CTypeSpec typeSpec)
+    typeSpec
     (u $ CDeclr
       (Just $ cIdent $ mangleName name)
       ((u $ CFunDeclr (Right (map cfunArg args, varargs)) []) : derivedDeclr)
@@ -41,7 +41,9 @@ cfunArg arg = cDecl (argType arg) (Just ([], argName arg)) Nothing
 
 attributesFromMeta :: [Metadata] -> [CAttr]
 attributesFromMeta (h : t) = case metaName h of
-  "inline"   -> (u $ CAttr (internalIdent "always_inline") []) : attributesFromMeta t
-  "noreturn" -> (u $ CAttr (internalIdent "noreturn") []) : attributesFromMeta t
-  _          -> attributesFromMeta t
+  "inline" ->
+    (u $ CAttr (internalIdent "always_inline") []) : attributesFromMeta t
+  "noreturn" ->
+    (u $ CAttr (internalIdent "noreturn") []) : attributesFromMeta t
+  _ -> attributesFromMeta t
 attributesFromMeta [] = []

--- a/src/Kit/Compiler/Ir/FindUnderlyingType.hs
+++ b/src/Kit/Compiler/Ir/FindUnderlyingType.hs
@@ -22,7 +22,7 @@ findUnderlyingType
 findUnderlyingType ctx mod pos t = _findUnderlyingType ctx mod pos [] t
 _findUnderlyingType ctx mod pos stack t = do
   tctx <- newTypeContext []
-  t <- mapType (follow ctx tctx) t
+  t    <- mapType (follow ctx tctx) t
   let r x = _findUnderlyingType ctx mod pos (x : stack) x
   when (length stack > 256) $ throwk $ InternalError
     ("Maximum recursion depth in findUnderlyingType exceeded; " ++ show stack)
@@ -55,7 +55,10 @@ _findUnderlyingType ctx mod pos stack t = do
     -- TypeEnumConstructor TypePath ConcreteArgs
     -- TypeRange
     -- TypeTraitPointer TypePath
-    TypeArray    t s        -> do
+    TypeConst t             -> do
+      t <- r t
+      return $ BasicTypeConst t
+    TypeArray t s -> do
       t <- r t
       return $ CArray t (if s == 0 then Nothing else Just s)
     TypeInstance (["kit", "common"], "CArray") [t, s] -> do

--- a/src/Kit/Compiler/Passes/IncludeCModules.hs
+++ b/src/Kit/Compiler/Passes/IncludeCModules.hs
@@ -255,6 +255,7 @@ _parseDeclSpec modPath (h : t) width signed float = case h of
      ]
     )
   _ -> _parseDeclSpec modPath t width signed float
+_parseDeclSpec modPath [] 0     False  _     = (TypeBasicType BasicTypeCUint)
 _parseDeclSpec modPath [] 0     _      _     = (TypeBasicType BasicTypeUnknown)
 _parseDeclSpec modPath [] width signed float = if float
   then TypeBasicType $ BasicTypeFloat width

--- a/src/Kit/Parser/Parser.y
+++ b/src/Kit/Parser/Parser.y
@@ -396,7 +396,7 @@ TypeParams_ :: {[TypeParam (Maybe TypeSpec)]}
 TypeParam :: {TypeParam (Maybe TypeSpec)}
   : TypeParam '=' TypeSpec {$1 {typeParamDefault = Just $ Just $ fst $3}}
   | upper_identifier TypeConstraints {(makeTypeParam (extract_upper_identifier $1)) {constraints = map Just $2, typeParamIsConstant = False, typeParamPos = snd $1}}
-  | macro_identifier {(makeTypeParam (s_drop 1 $ extract_macro_identifier $1)) {constraints = [], typeParamIsConstant = True, typeParamPos = snd $1}}
+  | macro_identifier {(makeTypeParam (extract_macro_identifier $1)) {constraints = [], typeParamIsConstant = True, typeParamPos = snd $1}}
 
 TypeSpecParams :: {([TypeSpec], Span)}
   : {([], NoPos)}

--- a/std/kit/common.kit
+++ b/std/kit/common.kit
@@ -11,6 +11,8 @@
 }
 #[builtin] abstract Box[T];
 
+#[builtin] #[promote] abstract Const[T]: T;
+
 #[promote] abstract ReadOnly[T]: T {
     rules {
         ($this.$field = $e) => {1 / 0;}

--- a/std/kit/utils.kit
+++ b/std/kit/utils.kit
@@ -15,7 +15,7 @@ rules FormatString {
     // println(s.toString());
 // }
 
-#[noreturn] function panic(msg: CString) {
+#[noreturn] function panic(msg: Const[CString]) {
     printf("%s\n", msg);
     exit(1);
 }

--- a/tests/Kit/Compiler/Passes/IncludeCModulesSpec.hs
+++ b/tests/Kit/Compiler/Passes/IncludeCModulesSpec.hs
@@ -1,6 +1,8 @@
 module Kit.Compiler.Passes.IncludeCModulesSpec where
 
 import Control.Monad
+import Data.Maybe
+import Language.C
 import Test.Hspec
 import Test.QuickCheck
 import Kit.Ast
@@ -46,7 +48,21 @@ spec = do
       ]
       (\t ->
         it ("Parses C specifiers into " ++ show t)
-          $          parseType [] (fst $ ctype t) []
+          $          parseType
+                       []
+                       (fst
+                         (let (a, b) = ctype t
+                          in  ( catMaybes $ map
+                                (\x -> case x of
+                                  CTypeSpec x -> Just x
+                                  _           -> Nothing
+                                )
+                                a
+                              , b
+                              )
+                         )
+                       )
+                       []
           `shouldBe` TypeBasicType t
       )
 
@@ -147,6 +163,10 @@ spec = do
           ]
           False
           []
+        )
+      , ( "Parses 'unsigned' by itself"
+        , "just_unsigned"
+        , (TypeBasicType $ BasicTypeCUint)
         )
       ]
       (\(label, name, ct) -> it label $ do

--- a/tests/Kit/Compiler/Passes/test_header.h
+++ b/tests/Kit/Compiler/Passes/test_header.h
@@ -35,6 +35,7 @@ short *pointer_var1;
 short **pointer_var2;
 short (*pointer_var3)(short arg1);
 void (*func_pointer)(void);
+unsigned just_unsigned;
 
 // functions
 

--- a/tests/compile-src/identifiers.kit
+++ b/tests/compile-src/identifiers.kit
@@ -8,7 +8,7 @@ function main() {
 
     // the function signature differs, so this will fail to compile unless
     // it resolve to the real builtin panic function
-    var x: function (CString) -> Void = kit.utils.panic;
+    var x: function (Const[CString]) -> Void = kit.utils.panic;
 
     var l1: List[Int] = kit.list.List.Empty;
     var l2: kit.list.List[Int] = kit.list.List[Int].Empty;

